File: go/net.go
package aun

import (
	"context"
	"crypto/tls"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	_ "modernc.org/sqlite"
)

// DnsResilientNet 统一 DNS 容灾网络层。
// 使用独立的 SQLite 文件 `{aun_path}/dns_cache.db` 持久化 DNS 缓存。
type DnsResilientNet struct {
	db        *sql.DB
	mu        sync.Mutex
	verifySSL bool
}

// NewDnsResilientNet 创建 DNS 容灾网络层
func NewDnsResilientNet(aunPath string, verifySSL bool) *DnsResilientNet {
	n := &DnsResilientNet{verifySSL: verifySSL}
	if aunPath != "" {
		dbPath := aunPath + "/dns_cache.db"
		db, err := sql.Open("sqlite", dbPath)
		if err == nil {
			_, _ = db.Exec(`CREATE TABLE IF NOT EXISTS dns_cache (
				hostname TEXT PRIMARY KEY,
				ip TEXT NOT NULL,
				port INTEGER NOT NULL DEFAULT 443,
				updated_at INTEGER NOT NULL
			)`)
			_, _ = db.Exec("PRAGMA journal_mode=WAL")
			n.db = db
		}
	}
	return n
}

// Close 关闭数据库连接
func (n *DnsResilientNet) Close() {
	n.mu.Lock()
	defer n.mu.Unlock()
	if n.db == nil {
		return
	}
	_ = n.db.Close()
	n.db = nil
}

func (n *DnsResilientNet) saveDNSCache(hostname, ip string, port int) {
	if n.db == nil || hostname == "" || ip == "" {
		return
	}
	n.mu.Lock()
	defer n.mu.Unlock()
	_, _ = n.db.Exec(
		"INSERT OR REPLACE INTO dns_cache (hostname, ip, port, updated_at) VALUES (?, ?, ?, ?)",
		hostname, ip, port, time.Now().Unix(),
	)
}

func (n *DnsResilientNet) loadDNSCache(hostname string) (ip string, port int, ok bool) {
	if n.db == nil || hostname == "" {
		return "", 0, false
	}
	n.mu.Lock()
	defer n.mu.Unlock()
	row := n.db.QueryRow("SELECT ip, port FROM dns_cache WHERE hostname = ?", hostname)
	if err := row.Scan(&ip, &port); err != nil {
		return "", 0, false
	}
	return ip, port, true
}

func (n *DnsResilientNet) refreshDNSCacheAfterSuccess(rawURL string) {
	hostname, port := parseHostPort(rawURL)
	if hostname == "" {
		return
	}
	ip := resolveIP(hostname)
	if ip != "" {
		n.saveDNSCache(hostname, ip, port)
	}
}

// HTTPGet DNS 容灾 HTTP GET
func (n *DnsResilientNet) HTTPGet(ctx context.Context, targetURL string, timeout time.Duration) ([]byte, error) {
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	hostname, port := parseHostPort(targetURL)

	// 1. 正常请求（域名）
	body, err := n.doHTTPGet(ctx, targetURL, "", timeout)
	if err == nil {
		n.refreshDNSCacheAfterSuccess(targetURL)
		return body, nil
	}
	if !isDNSError(err) {
		return nil, err
	}

	// 2. DNS 失败 → fallback 到缓存 IP
	cachedIP, cachedPort, ok := n.loadDNSCache(hostname)
	if !ok {
		return nil, fmt.Errorf("DNS failed and no cached IP for %s: %w", hostname, err)
	}
	_ = cachedPort
	ipURL := replaceHostWithIP(targetURL, cachedIP, port)
	return n.doHTTPGet(ctx, ipURL, hostname, timeout)
}

// HTTPGetJSON DNS 容灾 HTTP GET JSON
func (n *DnsResilientNet) HTTPGetJSON(ctx context.Context, targetURL string, timeout time.Duration) (map[string]any, error) {
	body, err := n.HTTPGet(ctx, targetURL, timeout)
	if err != nil {
		return nil, err
	}
	var payload map[string]any
	if err := json.Unmarshal(body, &payload); err != nil {
		return nil, fmt.Errorf("invalid JSON from %s: %w", targetURL, err)
	}
	return payload, nil
}

// HTTPGetOK DNS 容灾 HTTP GET 状态检查
func (n *DnsResilientNet) HTTPGetOK(ctx context.Context, targetURL string, timeout time.Duration) bool {
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	hostname, port := parseHostPort(targetURL)

	ok, err := n.doHTTPGetOK(ctx, targetURL, "", timeout)
	if err == nil && ok {
		n.refreshDNSCacheAfterSuccess(targetURL)
		return true
	}
	if err != nil && !isDNSError(err) {
		return false
	}

	cachedIP, _, cached := n.loadDNSCache(hostname)
	if !cached {
		return false
	}
	ipURL := replaceHostWithIP(targetURL, cachedIP, port)
	ok, _ = n.doHTTPGetOK(ctx, ipURL, hostname, timeout)
	return ok
}

func (n *DnsResilientNet) httpClient(serverName string) *http.Client {
	transport := &http.Transport{}
	if !n.verifySSL {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	} else if serverName != "" {
		transport.TLSClientConfig = &tls.Config{ServerName: serverName}
	}
	return &http.Client{Transport: transport}
}

func (n *DnsResilientNet) doHTTPGet(ctx context.Context, targetURL, sniHostname string, timeout time.Duration) ([]byte, error) {
	reqCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, targetURL, nil)
	if err != nil {
		return nil, err
	}
	if sniHostname != "" {
		req.Host = sniHostname
	}

	client := n.httpClient(sniHostname)
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d from %s", resp.StatusCode, targetURL)
	}
	return io.ReadAll(resp.Body)
}

func (n *DnsResilientNet) doHTTPGetOK(ctx context.Context, targetURL, sniHostname string, timeout time.Duration) (bool, error) {
	reqCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, targetURL, nil)
	if err != nil {
		return false, err
	}
	if sniHostname != "" {
		req.Host = sniHostname
	}

	client := n.httpClient(sniHostname)
	resp, err := client.Do(req)
	if err != nil {
		return false, err
	}
	resp.Body.Close()
	return resp.StatusCode == http.StatusOK, nil
}

// ── 工具函数 ──────────────────────────────────────────────

func isDNSError(err error) bool {
	if err == nil {
		return false
	}
	var dnsErr *net.DNSError
	if errors.As(err, &dnsErr) {
		return true
	}
	msg := strings.ToLower(err.Error())
	return strings.Contains(msg, "no such host") ||
		strings.Contains(msg, "name or service not known") ||
		strings.Contains(msg, "server misbehaving") ||
		strings.Contains(msg, "i/o timeout") && strings.Contains(msg, "lookup")
}

func parseHostPort(rawURL string) (string, int) {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return "", 443
	}
	hostname := parsed.Hostname()
	port := 443
	if parsed.Port() != "" {
		fmt.Sscanf(parsed.Port(), "%d", &port)
	} else if parsed.Scheme == "http" || parsed.Scheme == "ws" {
		port = 80
	}
	return hostname, port
}

func replaceHostWithIP(rawURL, ip string, port int) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL
	}
	hostPart := ip
	if strings.Contains(ip, ":") {
		hostPart = "[" + ip + "]"
	}
	if parsed.Port() != "" {
		parsed.Host = fmt.Sprintf("%s:%s", hostPart, parsed.Port())
	} else {
		parsed.Host = hostPart
	}
	return parsed.String()
}

func resolveIP(hostname string) string {
	ips, err := net.LookupHost(hostname)
	if err != nil || len(ips) == 0 {
		return ""
	}
	return ips[0]
}

File: ts/src/net.ts
/**
 * DNS-Resilient 网络层。
 *
 * 所有 HTTP/WebSocket 请求统一经过此模块，提供 DNS 容灾能力：
 * - 正常走域名（利用系统 DNS 缓存和 CDN 调度）
 * - 连接成功后刷新 DNS→IP 映射到 SQLite
 * - DNS 失败时 fallback 到持久化的最后一次成功 IP
 * - IP 直连时设置 TLS SNI + Host header
 */
import * as http from 'node:http';
import * as https from 'node:https';
import * as dns from 'node:dns';
import * as tls from 'node:tls';
import type { ModuleLogger } from './logger.js';

const _noopLogger: ModuleLogger = { error: () => {}, warn: () => {}, info: () => {}, debug: () => {} };

export interface DnsCacheStore {
  saveDnsCache(hostname: string, ip: string, port: number): void;
  loadDnsCache(hostname: string): { ip: string; port: number } | null;
}

function isDnsError(err: unknown): boolean {
  if (!err || typeof err !== 'object') return false;
  const code = (err as { code?: string }).code;
  if (code === 'ENOTFOUND' || code === 'EAI_AGAIN' || code === 'EAI_FAIL' || code === 'EAI_NONAME') {
    return true;
  }
  const msg = String((err as Error).message ?? '').toLowerCase();
  return msg.includes('getaddrinfo') && (msg.includes('enotfound') || msg.includes('eai_again'));
}

function parseHostPort(url: string): { hostname: string; port: number } {
  const parsed = new URL(url);
  const hostname = parsed.hostname;
  let port = parsed.port ? Number(parsed.port) : (parsed.protocol === 'https:' || parsed.protocol === 'wss:' ? 443 : 80);
  return { hostname, port };
}

function replaceHostWithIP(url: string, ip: string): string {
  const parsed = new URL(url);
  parsed.hostname = ip.includes(':') ? `[${ip}]` : ip;
  return parsed.toString();
}

function resolveIP(hostname: string): Promise<string | null> {
  return new Promise((resolve) => {
    dns.lookup(hostname, (err, address) => {
      resolve(err ? null : address);
    });
  });
}

export class DnsResilientNet {
  private _store: DnsCacheStore | null;
  private _verifySsl: boolean;
  private _logger: ModuleLogger;

  constructor(opts?: { store?: DnsCacheStore; verifySsl?: boolean; logger?: ModuleLogger }) {
    this._store = opts?.store ?? null;
    this._verifySsl = opts?.verifySsl ?? true;
    this._logger = opts?.logger ?? _noopLogger;
  }

  private _saveDnsCache(hostname: string, ip: string, port: number): void {
    if (!this._store || !hostname || !ip) return;
    try {
      this._store.saveDnsCache(hostname, ip, port);
    } catch (exc) {
      this._logger.debug(`dns cache save failed: ${exc}`);
    }
  }

  private _loadDnsCache(hostname: string): { ip: string; port: number } | null {
    if (!this._store || !hostname) return null;
    try {
      return this._store.loadDnsCache(hostname);
    } catch (exc) {
      this._logger.debug(`dns cache load failed: ${exc}`);
      return null;
    }
  }

  private async _refreshDnsCacheAfterSuccess(url: string): Promise<void> {
    const { hostname, port } = parseHostPort(url);
    if (!hostname) return;
    const ip = await resolveIP(hostname);
    if (ip) this._saveDnsCache(hostname, ip, port);
  }

  async httpGet(url: string, timeout = 5_000): Promise<Buffer> {
    const { hostname, port } = parseHostPort(url);

    try {
      const data = await this._doHttpGet(url, undefined, timeout);
      void this._refreshDnsCacheAfterSuccess(url);
      return data;
    } catch (exc) {
      if (!isDnsError(exc)) throw exc;
    }

    this._logger.debug(`DNS failed for ${hostname}, trying cached IP`);
    const cached = this._loadDnsCache(hostname);
    if (!cached) {
      const err = new Error(`DNS failed and no cached IP for ${hostname}`);
      (err as any).code = 'ENOTFOUND';
      throw err;
    }

    const ipUrl = replaceHostWithIP(url, cached.ip);
    return this._doHttpGet(ipUrl, hostname, timeout);
  }

  async httpGetJson(url: string, timeout = 5_000): Promise<Record<string, unknown>> {
    const data = await this.httpGet(url, timeout);
    return JSON.parse(data.toString('utf-8')) as Record<string, unknown>;
  }

  async httpGetText(url: string, timeout = 5_000): Promise<string> {
    const data = await this.httpGet(url, timeout);
    return data.toString('utf-8');
  }

  async httpGetOk(url: string, timeout = 5_000): Promise<boolean> {
    const { hostname, port } = parseHostPort(url);

    try {
      const ok = await this._doHttpGetOk(url, undefined, timeout);
      if (ok) void this._refreshDnsCacheAfterSuccess(url);
      return ok;
    } catch (exc) {
      if (!isDnsError(exc)) return false;
    }

    const cached = this._loadDnsCache(hostname);
    if (!cached) return false;

    const ipUrl = replaceHostWithIP(url, cached.ip);
    try {
      return await this._doHttpGetOk(ipUrl, hostname, timeout);
    } catch {
      return false;
    }
  }

  /** 导出供 WebSocket 连接层使用 */
  get isDnsError() { return isDnsError; }
  get parseHostPort() { return parseHostPort; }
  get replaceHostWithIP() { return replaceHostWithIP; }

  loadDnsCache(hostname: string) { return this._loadDnsCache(hostname); }
  refreshDnsCacheAfterSuccess(url: string) { return this._refreshDnsCacheAfterSuccess(url); }

  private _doHttpGet(url: string, sniHostname: string | undefined, timeout: number): Promise<Buffer> {
    return new Promise((resolve, reject) => {
      const parsed = new URL(url);
      const mod = parsed.protocol === 'https:' ? https : http;
      const options: https.RequestOptions = { timeout };
      if (!this._verifySsl) {
        options.rejectUnauthorized = false;
      }
      if (sniHostname) {
        options.headers = { Host: sniHostname };
        options.servername = sniHostname;
      }

      const req = mod.get(url, options, (res: http.IncomingMessage) => {
        if (res.statusCode && (res.statusCode < 200 || res.statusCode >= 300)) {
          reject(new Error(`HTTP ${res.statusCode} from ${url}`));
          res.resume();
          return;
        }
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => resolve(Buffer.concat(chunks)));
        res.on('error', reject);
      });
      req.on('error', reject);
      req.on('timeout', () => { req.destroy(); reject(new Error(`timeout fetching ${url}`)); });
    });
  }

  private _doHttpGetOk(url: string, sniHostname: string | undefined, timeout: number): Promise<boolean> {
    return new Promise((resolve, reject) => {
      const parsed = new URL(url);
      const mod = parsed.protocol === 'https:' ? https : http;
      const options: https.RequestOptions = { method: 'GET', timeout };
      if (!this._verifySsl) options.rejectUnauthorized = false;
      if (sniHostname) {
        options.headers = { Host: sniHostname };
        options.servername = sniHostname;
      }

      const req = mod.request(url, options, (res) => {
        res.resume();
        resolve(res.statusCode === 200);
      });
      req.on('error', (err) => {
        if (isDnsError(err)) reject(err);
        else resolve(false);
      });
      req.on('timeout', () => { req.destroy(); resolve(false); });
      req.end();
    });
  }
}

export { isDnsError, parseHostPort, replaceHostWithIP };
